src/main/resources/avro/readmethods.avdl

```diff
@@ -14,8 +14,10 @@ to `ReferenceSet`s containing that same `Reference`. If no reference is
 specified, all `ReadGroup`s must be aligned to the same `ReferenceSet`.
 */
 record SearchReadsRequest {
-  /** If nonempty, restrict this query to reads within the given readgroups. */
-  array<string> readGroupIds = [];
+  /**
+  The ReadGroup to search.
+  */
+  string readGroupId;
```

Comment:
We have to feel free to break compatibility when necessary.

The API should be versioned, but 0.5 is way too immature to be
anything set in stone.  We have a weak conformance suite and
virtually nothing real (at least that we have seen) has been
built with it.  If we don't validate by building real
applications while the API is being design, we will end up with
a mess.

The world will have to live with this for probably 20 years.  We
can't do a design-by-committee, through over the wall
specification and expect it to be right. This is an absolute
recipe for failure.  We have to build real applications and
iterate.

This is why the reference service is so important to GA4GH's
success.

Mark

Jerome Kelleher <notifications@github.com> writes:
> Thanks for all the comments, I think this is a very useful conversation. Some
> responses:
> 
>   • @dglazer I appreciate your point about this not being a green-field site, I
>     really do. Breaking compatibility is a big deal, and not something we
>     should take lightly. On the other hand though, this is a pre-1.0 API, and
>     we must have the freedom to make some changes if we are to develop it. The
>     changes here are simplifications, and as you say, if we don't make these
>     now it'll be very difficult to make them later. Users can be shielded from
>     these changes by the client-side libraries quite easily, so I don't think
>     any serious existing applications need be affected at all. For example,
>     api-client-java can detect which version of the API a server is running,
>     and send sequential requests when existing code uses the multi-container
>     search semantics.
> 
>   • @lh3 This is a good example. If we have the need to merge reads from 1000
>     ReadGroupSets, and the number of reads per ReadGroupSet is often zero or
>     much less than the page size, then this would be a compelling performance
>     argument in favour of doing the merge on the server side. I don't want to
>     be dogmatic about this, and if this is a real requirement that can't be
>     worked around then I'm happy to keep searches across ReadGroupSets.
>     However, why would we want to search across 1000 ReadGroupSets? If we have
>     reads for 1000 related samples, why not keep them all in a single
>     ReadGroupSet? I'm happy to add the ability to search over a single
>     ReadGroupSet by adding a readGroupSetId field as @calbach suggested at the
>     start of this thread. Would this satisfy the most common requirements? As 
>     @richarddurbin pointed out, we can't solve this problem on the server side
>     in general in any case, since we might need to merge reads from different
>     servers.
> 
>   • @richarddurbin This is a very important question I think. What is the real
>     API: the low-level transport layer of HTTP and JSON, or the API that
>     developers actually interact with in their language of choice when writing
>     applications? As a client-side developer (in say C, Java, Python, R or
>     JavaScript), I don't want to know about HTTP, JSON, or paging any more than
>     I want to know what my TCP packet size is. What I want is a well-defined
>     client-side API in my language with documentation written specifically for
>     that language. A client-side API will look very different in C and R, and
>     they must have native documentation if developers are going to write
>     applications. Unless we provide excellent client libraries that make
>     writing applications substantially easier than using existing tools, there
>     will be no significant reason for developers to adopt the GA4GH API.
> 
> —
> Reply to this email directly or view it on GitHub.*
>